runtime/runtime_events.c

```diff
@@ -404,20 +396,37 @@ static void runtime_events_create_raw(void) {
   }
 }
 
+/* create/teardown STWs
+
+   The STW API does have an enter barrier before the handler code is
+   run, however, the enter barrier itself calls the runtime events API
+   after arrival. Thus, the barrier in the STWs below is needed both
+   to ensure that all domains have actually reached the handler before
+   we start/stop (and are not calling the runtime events API from the
+   STW code), and of course to ensure that the setup/teardown is
+   observed by all domains returning from the STW. */
+
 /* Stop the world section which calls [runtime_events_create_raw], used when we
    can't be sure there is only a single domain running. */
 static void
 stw_create_runtime_events(caml_domain_state *domain_state, void *data,
-                              int num_participating,
-                              caml_domain_state **participating_domains) {
+                          int num_participating,
+                          caml_domain_state **participating_domains) {
   /* Everyone must be stopped for starting and stopping runtime_events */
-  caml_global_barrier();
-
-  /* Only do this on one domain */
-  if (participating_domains[0] == domain_state) {
+  Caml_global_barrier_if_final(num_participating) {
     runtime_events_create_raw();
   }
-  caml_global_barrier();
+}
+
+/* Stop-the-world which calls the teardown code */
+static void
+stw_teardown_runtime_events(caml_domain_state *domain_state,
+                            void *remove_file_data, int num_participating,
+                            caml_domain_state **participating_domains) {
+  Caml_global_barrier_if_final(num_participating) {
```

Comment:
I have the same difficulty to understand this barrier as the one in `stw_create_runtime_events` above.

Reply:
Resolved (see related comment thread).